File: Source/Notifications/Push notifications/Notification Types/ZMLocalNotification.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


@import WireSystem;
@import WireUtilities;
@import WireTransport;
@import WireDataModel;

#import "ZMLocalNotification.h"

#import "ZMLocalNotificationLocalization.h"
#import "UILocalNotification+StringProcessing.h"
#import "ZMUserSession+UserNotificationCategories.h"
#import "WireSyncEngine/WireSyncEngine-Swift.h"
#import "UILocalNotification+UserInfo.h"


NSString * const ZMLocalNotificationConversationObjectURLKey = @"conversationObjectURLString";
NSString * const ZMLocalNotificationUserInfoSenderKey = @"senderUUID";
NSString * const ZMLocalNotificationUserInfoNonceKey = @"nonce";

static NSString * const FailedMessageInGroupConversationText = @"failed.message.group";
static NSString * const FailedMessageInOneOnOneConversationText = @"failed.message.oneonone";

NSString * const ZMPushStringDefault = @"default";
NSString * const ZMPushStringEphemeral = @"ephemeral";


// These are the "base" keys for messages. We append to these for the specific case.
//
//
//
// 1 user, 1 conversation, 1 string
// %1$@    %2$@            %3$@
NSString *const ZMPushStringMessageAdd = @"add.message"; // "[senderName] in [conversationName] - [messageText]"
NSString *const ZMPushStringImageAdd = @"add.image"; // "[senderName] shared a picture in [conversationName]"
NSString *const ZMPushStringVideoAdd = @"add.video"; // "[senderName] shared a video in [conversationName]"
NSString *const ZMPushStringAudioAdd = @"add.audio"; // "[senderName] shared an audio message in [conversationName]"
NSString *const ZMPushStringFileAdd = @"add.file"; // "[senderName] shared a file in [conversationName]"
NSString *const ZMPushStringLocationAdd = @"add.location"; // "[senderName] shared a location in [conversationName]"
NSString *const ZMPushStringUnknownAdd = @"add.unknown"; // "[senderName] sent a message in [conversationName]"
NSString *const ZMPushStringMessageAddMany = @"add.message.many"; // "x new messages in [conversationName] / from [senderName]"

/// 2 users, 1 conversation
NSString *const ZMPushStringMemberJoin = @"member.join";
// "[senderName] added you / [userName] to [conversationName]"
NSString *const ZMPushStringMemberLeave = @"member.leave";
// "[senderName] removed you / [userName] from [conversationName]"


NSString *const ZMPushStringMemberJoinMany = @"member.join.many";
// "[senderName] added people to [conversationName]"
NSString *const ZMPushStringMemberLeaveMany = @"member.leave.many";
// "[senderName] removed people from [conversationName]"

NSString *const ZMPushStringKnock = @"knock"; // "[senderName] pinged you x times in [conversationName]" // "x pings in
NSString *const ZMPushStringReaction = @"reaction"; // "[senderName] [emoji] your message in [conversationName]"

NSString *const ZMPushStringVideoCallStarts = @"call.started.video"; // "[senderName] wants to talk"
NSString *const ZMPushStringCallStarts = @"call.started"; // "[senderName] wants to talk"
NSString *const ZMPushStringCallMissed = @"call.missed"; // "[senderName] called you x times"
NSString *const ZMPushStringCallMissedMany = @"call.missed.many"; // "You have x missed calls in a conversation"

NSString *const ZMPushStringConnectionRequest = @"connection.request"; // "[senderName] wants to connect: [messageText]"
NSString *const ZMPushStringConnectionAccepted = @"connection.accepted"; // "[senderName] accepted your connection request"

NSString *const ZMPushStringConversationCreate = @"conversation.create";
NSString *const ZMPushStringNewConnection = @"new_user";

#pragma mark - ZMLocalNotification


@interface ZMLocalNotification ()

@end


@implementation ZMLocalNotification


- (instancetype)initWithConversationID:(NSUUID *)conversationID
{
    self = [super init];
    if (self) {
        _conversationID = conversationID;
    }
    return self;
}

- (NSArray<UILocalNotification *> *)uiNotifications
{
    return  @[];
}

@end



#pragma mark - ZMLocalNotificationForExpiredMessage


@implementation ZMLocalNotificationForExpiredMessage



- (instancetype)initWithExpiredMessage:(ZMMessage *)message
{
    self = [super initWithConversationID:message.conversation.remoteIdentifier];
    if(self) {
        _message = message;
        UILocalNotification *note = [[UILocalNotification alloc] init];
        [self createBodyForConversation:message.conversation notification:note];
        _uiNotification = note;
    }
    return self;
}

- (NSArray<UILocalNotification *> *)uiNotifications
{
    return @[self.uiNotification];
}

- (instancetype)initWithConversation:(ZMConversation *)conversation
{
    self = [super initWithConversationID:conversation.remoteIdentifier];
    if(self) {
        UILocalNotification *note = [[UILocalNotification alloc] init];
        [self createBodyForConversation:conversation notification:note];
        _uiNotification = note;
    }
    return self;
}

- (void)createBodyForConversation:(ZMConversation *)conversation notification:(UILocalNotification *)notification
{
    if(conversation.conversationType == ZMConversationTypeGroup) {
        notification.alertBody = [FailedMessageInGroupConversationText localizedStringWithConversation:conversation count:nil];
    }
    else {
        notification.alertBody = [FailedMessageInOneOnOneConversationText localizedStringWithUser:conversation.connectedUser count:nil];
    }
    [notification setupUserInfo:conversation forEvent:nil];
}

@end

File: Source/Notifications/Push notifications/Notification Types/ZMLocalNotification.swift
//
// Wire
// Copyright (C) 2017 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit
import UserNotifications
import WireSystem
import WireUtilities
import WireTransport
import WireDataModel


/// Defines the various types of local notifications, some of which
/// have associated subtypes.
///
public enum LocalNotificationType {
    case event(LocalNotificationEventType)
    case calling(CallState)
    case message(LocalNotificationContentType)
    case failedMessage
}

/// A notification builder provides the main components used to configure
/// a local notification. 
///
protocol NotificationBuilder {
    var notificationType : LocalNotificationType { get }
    func shouldCreateNotification() -> Bool
    func titleText() -> String?
    func bodyText() -> String
    func userInfo() -> [AnyHashable: Any]?
}


/// This class encapsulates all the data necessary to produce a local
/// notification. It configures and formats the textual content for
/// various notification types (message, calling, etc.) and includes
/// information regarding the conversation, sender, and team name.
///
open class ZMLocalNotification: NSObject {
    
    public let type: LocalNotificationType
    public var title: String?
    public var body: String
    public var category: String?
    public var soundName: String?
    public var userInfo: [AnyHashable: Any]?
        
    public var selfUserID: UUID? { return uuid(for: SelfUserIDStringKey) }
    public var senderID: UUID? { return uuid(for: SenderIDStringKey) }
    public var messageNonce: UUID? { return uuid(for: MessageNonceIDStringKey) }
    public var conversationID: UUID? { return uuid(for: ConversationIDStringKey) }
    
    init?(conversation: ZMConversation?, builder: NotificationBuilder) {
        guard builder.shouldCreateNotification() else { return nil }
        self.type = builder.notificationType
        self.title = builder.titleText()
        self.body = builder.bodyText().escapingPercentageSymbols()
        self.category = builder.notificationType.category
        self.soundName = builder.notificationType.soundName
        self.userInfo = builder.userInfo()
    }
    
    /// Returns a configured concrete UILocalNotification object.
    ///
    public lazy var uiLocalNotification: UILocalNotification = {
        let note = UILocalNotification()
        
        let candidateTitle = self.title
        let candidateBody = self.body
        
        if #available(iOS 10, *) {
            note.alertTitle = candidateTitle
            note.alertBody = candidateBody
        }
        else {
            // on iOS 9, the alert title is only visible in the notification center, so we
            // display all info in the body
            if let title = candidateTitle {
                note.alertBody = "\(title)\n\(candidateBody)"
            } else {
                note.alertBody = candidateBody
            }
        }
        
        note.category = self.category
        note.soundName = self.soundName
        note.userInfo = self.userInfo
        return note
    }()
    
    /// Returns true if it is a calling notification, else false.
    ///
    public var isCallingNotification: Bool {
        switch type {
        case .calling: return true
        default: return false
        }
    }
    
    /// Returns true if it is a ephemeral notification, else false.
    ///
    public var isEphemeral: Bool {
        switch type {
        case .message(let contentType):
            if case .ephemeral = contentType {
                return true
            } else {
                return false
            }
        default:
            return false
        }
    }
    
    /// Returns the UUID for the given key from the user info if it exists, else
    /// nil.
    ///
    private func uuid(for key: String) -> UUID? {
        guard let uuidString = userInfo?[key] as? String else { return nil }
        return UUID(uuidString: uuidString)
    }
    
    public func conversation(in moc: NSManagedObjectContext) -> ZMConversation? {
        guard let uuid = conversationID else { return nil }
        return ZMConversation(remoteID: uuid, createIfNeeded: false, in: moc)
    }
    
    public func sender(in moc: NSManagedObjectContext) -> ZMUser? {
        guard let uuid = senderID else { return nil }
        return ZMUser(remoteID: uuid, createIfNeeded: false, in: moc)
    }
}
